chain/client/src/client.rs

```diff
@@ -229,6 +242,91 @@ pub struct ProduceChunkResult {
     pub transactions_storage_proof: Option<PartialState>,
 }
 
+/// This keeps track of the numbef of new chunks seen in each shard since the block that was passed to new()
+/// This whole thing could be replaced with a much simpler function that just computes the numbef of new chunks
+/// in each shard from scratch every time we call it, but in the unlikely and unfortunate case where a shard
+/// hasn't had any chunks for a very long time, it would end up being a nontrivial inefficiency to do that
+/// every time run_catchup() is called
+pub struct NewChunkTracker {
+    last_checked_hash: CryptoHash,
+    last_checked_height: BlockHeight,
+    num_new_chunks: HashMap<ShardId, usize>,
+    sync_hash: Option<CryptoHash>,
+}
+
+impl NewChunkTracker {
+    fn new(
+        first_block_hash: CryptoHash,
+        first_block_height: BlockHeight,
+        shard_ids: &[ShardId],
+    ) -> Self {
+        Self {
+            last_checked_hash: first_block_hash,
+            last_checked_height: first_block_height,
+            num_new_chunks: shard_ids.iter().map(|shard_id| (*shard_id, 0)).collect(),
+            sync_hash: None,
+        }
+    }
+
+    fn record_new_chunks(&mut self, header: &BlockHeader) -> Result<bool, Error> {
+        let mut done = true;
+        for (shard_id, num_new_chunks) in self.num_new_chunks.iter_mut() {
+            let included = match header.chunk_mask().get(*shard_id as usize) {
+                Some(i) => *i,
+                None => {
+                    return Err(Error::Other(format!(
+                        "can't get shard {} in chunk mask for block {}",
+                        shard_id,
+                        header.hash()
+                    )));
+                }
+            };
+            if included {
+                *num_new_chunks += 1;
+            }
+            if *num_new_chunks < 2 {
+                done = false;
+            }
+        }
+        self.last_checked_hash = *header.hash();
+        self.last_checked_height = header.height();
+        Ok(done)
+    }
+
+    fn find_sync_hash(&mut self, chain: &Chain) -> Result<Option<CryptoHash>, Error> {
```

Comment:
I think I saw another method with the same name. Is there any chance to merge those or make one a thin wrapper of the other?

Reply:
^ like in the above comment, I plan to delete `get_current_epoch_sync_hash` in another PR, so hopefully this will all just be in one place. If you think it should be done in this PR though, let me know